services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/modification/AbstractGradeChangeSubscriptionModificationHandler.java

```diff
@@ -157,4 +167,125 @@ protected SubscriptionReferencedItem readNewSubscriptionItemOfCustomType(
                         "Item reference type not supported. Root item reference type: %s. New item reference type: %s.",
                         subscription.getRootItemRefType(), request.getNewItemRefType()));
     }
+
+    /**
+     * Builds a {@link AddItemRequest} for the upgrade/downgrade product.
+     *
+     * <p>
+     * If the new product to upgrade/downgrade to has a required add-on, an {@link AddItemRequest}
+     * is created for the add-on too.
+     * </p>
+     *
+     * @param request The {@link ModifySubscriptionRequest} to hydrate. * @param contextInfo
+     *        Additional sandbox and multitenant info.
+     * @param flowName The name of the flow. Can be a value from
+     *        {@link DefaultSubscriptionActionFlow#DOWNGRADE} or
+     *        {@link DefaultSubscriptionActionFlow#UPGRADE}.
+     * @param contextInfo Additional sandbox and multitenant info.
+     * @return The {@link AddItemRequest} for the new product to upgrade/downgrade to.
+     */
+    protected AddItemRequest createAddItemRequestForGradeChangeProduct(
+            @lombok.NonNull ModifySubscriptionRequest request,
+            @lombok.NonNull String flowName,
+            @Nullable ContextInfo contextInfo) {
```

Comment:
For strings, we generally would consider a blank string to also be an illegal argument. So instead of just checking for null with lombok, we instead use Spring's `Assert.hasText` 

```suggestion
    protected AddItemRequest createAddItemRequestForGradeChangeProduct(
            @lombok.NonNull ModifySubscriptionRequest request,
            String flowName,
            @Nullable ContextInfo contextInfo) {
            Assert.hasText(flowName, "FlowName cannot be blank");
```